monai/handlers/classification_saver.py

```diff
@@ -86,12 +93,34 @@ def __call__(self, engine: Engine) -> None:
         Args:
             engine: Ignite Engine, it can be a trainer, validator or evaluator.
         """
-        _meta_data = self.batch_transform(engine.state.batch)
-        if Key.FILENAME_OR_OBJ in _meta_data:
-            # all gather filenames across ranks, only filenames are necessary
-            _meta_data = {Key.FILENAME_OR_OBJ: string_list_all_gather(_meta_data[Key.FILENAME_OR_OBJ])}
-        # all gather predictions across ranks
-        _engine_output = evenly_divisible_all_gather(self.output_transform(engine.state.output))
+        filenames = self.batch_transform(engine.state.batch).get(Key.FILENAME_OR_OBJ, None)
+        if filenames is not None:
+            self._filenames.extend(filenames)
+        outputs = self.output_transform(engine.state.output)
+        if outputs is not None:
+            self._outputs.append(outputs)
+
+    def _finalize(self, engine: Engine) -> None:
+        """
+        All gather classification results from ranks and save to CSV file.
+
+        Args:
+            engine: Ignite Engine, it can be a trainer, validator or evaluator.
+        """
+        outputs = evenly_divisible_all_gather(torch.cat(self._outputs, dim=0))
+        filenames = string_list_all_gather(self._filenames)
+        if len(filenames) == 0:
+            meta_dict = None
+        elif len(filenames) != len(outputs):
+            raise RuntimeError(f"filenames length: {len(filenames)} doesn't match outputs length: {len(outputs)}.")
+        else:
+            meta_dict = {Key.FILENAME_OR_OBJ: filenames}
 
+        # save to CSV file only in the expected rank
         if self._expected_rank:
-            self.saver.save_batch(_engine_output, _meta_data)
+            saver = CSVSaver(self.output_dir, self.filename, self.overwrite)
```

Comment:
`self.saver` is not needed? should be `self.filename` or `filenames`?

Reply:
deleted in the latest commit.
Thanks.